_posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md

```diff
@@ -0,0 +1,93 @@
+
+At first sight, it seems like a relatively simple task to make sure that this math is correct. It turns out that handling edge cases near the min and max durations, or when performing operations between very large and very small durations requires special attention, or even when crossing the boundary of the reference epoch. For example, the TAI reference epoch is 01 January 1900, so when subtracting one nanosecond from 01 January 1900 at midnight, the internal duration representation goes from `centuries: 0, nanoseconds: 0` to `centuries: -1, nanoseconds: 3_155_759_999_999_999_999`, as there are `3155759999999999999 + 1 = 3155760000000000000` nanoseconds in one century.
+
+With Kani, we can check for all permutations of a definition of a `Duration`, and ensure that the decomposition of a `Duration` into its parts of days, hours, minutes, seconds, milliseconds, microseconds, and nanoseconds, never causes any overflow, underflow, or other undefined behaviors. In hifitime, this is done simply by implementing `Arbitrary` for `Duration`, and calling the `decompose()` function on _any_ duration. This test is beautiful in its simplicity: small code footprint for mighty guarantees, such could be the motto of Kani!
+
+```rust
+#[cfg(kani)]
+impl Arbitrary for Duration {
+    #[inline(always)]
+    fn any() -> Self {
+        let centuries: i16 = kani::any();
+        let nanoseconds: u64 = kani::any();
+
+        Duration::from_parts(centuries, nanoseconds)
+    }
+}
+
+// (...)
+
+#[cfg(kani)]
+#[kani::proof]
+fn formal_duration_normalize_any() {
+    let dur: Duration = kani::any();
+    // Check that decompose never fails
+    dur.decompose();
+}
+```
+
+In the test above, if the call to `decompose` ever causes any overflow, underflow, division by zero, etc. then Kani will report an error and the exact value and bits inputted that caused the problem. This feedback is great, because one can just plugin those values into another test and debug it precisely. One will note that this test does not actually check the output value. That's for two reasons. First, testing values for all possible combinations would require re-implementing the same code that is in the `decompose` function making the test a tautology. Second, the purpose of the Hifitime Kani tests is to ensure that the aren't any unsound operations. Moreover, Hifitime has plenty of values that are explicitly tested for in the rest the tests.
```

Comment:
> That's for two reasons. First, testing values for all possible combinations would require re-implementing the same code that is in the `decompose` function making the test a tautology.

This sentence makes it seem like there would be no value in checking the output of the operation and that tautology checking is not that useful. However formal verification is effectively tautology checking ! So I think the message could be confusing to some. 

Could we instead say something like : there would be some value in checking that the output of the operation is as expected, if we had written a declarative, formal specification of the input-output relation of `decompose`. However this was not a goal in this project." 

And maybe save that for future work?.

We could also equivalence check this function against a second and independently developed equivalent.

Reply:
What do you think of the updated version?

> That's for two reasons. First, testing values would require developing either a formal specification or an independent implementation of the `decompose` function against which to check. Both options require quite a bit of effort, and the `decompose` function is really only used for displaying a duration in a human-readable format. Second, ...